### PR TITLE
[FLINK] [#2331] Add support to partition table by date type

### LINKED ISSUE #2331: [BUG][Flink] Caused by: java.lang.RuntimeException: Type not supported DATE
## Bug

- [ ] Spark
- [ ] Standalone
- [x] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Describe the problem
I know it's not actually a bug but I would like to mention my issue here.
When I am trying to using `Flink DeltaSink` to write a `DataStream<RowData>` of the following `RowType`, and partitioned by the column `effective_date` with type `DATE`, the exception is thrown.

```java
 public static final RowType ROW_TYPE = new RowType(Arrays.asList(
            new RowType.RowField("word", new VarCharType(VarCharType.MAX_LENGTH)),
            new RowType.RowField("frequency", new IntType()),
            new RowType.RowField("effective_date", new DateType())
    ));
```

```log
Caused by: java.lang.RuntimeException: Type not supported DATE
	at io.delta.flink.sink.internal.DeltaPartitionComputer$DeltaRowDataPartitionComputer.generatePartitionValues(DeltaPartitionComputer.java:103)
	at io.delta.flink.sink.internal.DeltaPartitionComputer$DeltaRowDataPartitionComputer.generatePartitionValues(DeltaPartitionComputer.java:43)
	at io.delta.flink.sink.internal.DeltaBucketAssigner.getBucketId(DeltaBucketAssigner.java:100)
	at io.delta.flink.sink.internal.DeltaBucketAssigner.getBucketId(DeltaBucketAssigner.java:87)
	at io.delta.flink.sink.internal.writer.DeltaWriter.write(DeltaWriter.java:288)
```	
	
I have checked the source at https://github.com/delta-io/connectors/blob/master/flink/src/main/java/io/delta/flink/sink/internal/DeltaPartitionComputer.java
and found out that currenly, flink delta connector only supports 5 Logical Types:

```java
if (staticPartitionSpec.containsKey(partitionKey)) {
    // We want the output partition values to be String's anyways, so no need
    // to parse/cast the staticPartitionSpec value
    partitionValues.put(partitionKey, staticPartitionSpec.get(partitionKey));
} else if (keyType.getTypeRoot() == LogicalTypeRoot.VARCHAR) {
    partitionValues.put(partitionKey, element.getString(keyIndex).toString());
} else if (keyType.getTypeRoot() == LogicalTypeRoot.INTEGER) {
    partitionValues.put(partitionKey, String.valueOf(element.getInt(keyIndex)));
} else if (keyType.getTypeRoot() == LogicalTypeRoot.BIGINT) {
    partitionValues.put(partitionKey, String.valueOf(element.getLong(keyIndex)));
} else if (keyType.getTypeRoot() == LogicalTypeRoot.SMALLINT) {
    partitionValues.put(partitionKey, String.valueOf(element.getShort(keyIndex)));
} else if (keyType.getTypeRoot() == LogicalTypeRoot.TINYINT) {
    partitionValues.put(partitionKey, String.valueOf(element.getByte(keyIndex)));
} else {
    throw new RuntimeException("Type not supported " + keyType.getTypeRoot());
}
```

### What I expected?

Why we do not utilize converters inside this package `org.apache.flink.table.data.conversion.*` to support more `Logical Type`,
for example, I could use the `DateDateConverter` and modify the code so that we can convert `INT` representation of a date to `yyyy-mm-dddd` representation?

The code below is just for demonstration:

```java
} else if (keyType.getTypeRoot() == LogicalTypeRoot.TINYINT) {
      partitionValues.put(partitionKey, String.valueOf(element.getByte(keyIndex)));
  } else if (keyType.getTypeRoot() == LogicalTypeRoot.DATE) {
      DateDateConverter converter = new DateDateConverter();
      String value = String.valueOf(converter.toExternal(element.getInt(keyIndex)));
      partitionValues.put(partitionKey, value); // now value is `yyyy-mm-dd` string
  } else {
      throw new RuntimeException("Type not supported " + keyType.getTypeRoot());
  }
  ```    

### Environment information

* Delta Lake version: 0.6.0 and 3.0.0
* Spark version: no
* Scala version: 2.12

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [ ] No. I cannot contribute a bug fix at this time.

### PR DESCRIPTION
**Pull request by tuliocavalcanti**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [x] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Adds support for table partition of DATE type for RowData elements.
Resolves #2331

## How was this patch tested?

Tests were updated to include the new type.
## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

**Review by andrelu — APPROVED** [review]

**Review by scottsand-db — CHANGES_REQUESTED** [review]
Looks great! Can you please add a simple test that writes multiple different date values?

If you need help or don't know how -- no worries, just let me know and I can take care of it!

**Comment by scottsand-db** [comment]
Hi @tuliocavalcanti - friendly follow up?

**Comment by tuliocavalcanti** [comment]
Hi Scott,
sorry for the delay, had a tight deadline to deliver the week.

could you please provide some date examples that you want tested?
I didn't do many testing because I'm using Flink's internal date converter `DateDateConverter` and expect it to be properly tested.

**Comment by scottsand-db** [comment]
Hi @tuliocavalcanti - I'd like an end to end test case that streams and writes different date values to a table. e.g. something as simple as start with 2024-09-01 and write values until 2024-10-01.

**Comment by tuliocavalcanti** [comment]
Sure thing, is there a another test case that I can use as guide to do so?
I'm not sure where to put this test, I'm assuming its not in this `TestDeltaBucketAssigner` file

**Comment by scottsand-db** [comment]
@tuliocavalcanti something like this?

```java
package io.delta.flink.sink;

import io.delta.flink.utils.DeltaTestUtils;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.data.RowData;
import org.apache.flink.table.data.util.DataFormatConverters;
import org.apache.flink.table.types.logical.DateType;
import org.apache.flink.table.types.logical.IntType;
import org.apache.flink.table.types.logical.RowType;
import org.apache.flink.table.types.utils.TypeConversions;
import org.apache.flink.types.Row;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.LocalDate;
import java.util.Arrays;

public class DeltaSinkSuite {

    @Test
    public void testWritePartitionedByDate(@TempDir Path tempDir) throws Exception {
        final String deltaTablePath = tempDir.toString();

        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        final RowType rowType = new RowType(Arrays.asList(
            new RowType.RowField("part1", new DateType()),
            new RowType.RowField("data1", new IntType())
        ));

        final DataFormatConverters.DataFormatConverter<RowData, Row> typeConverter =
            DataFormatConverters.getConverterForDataType(
                TypeConversions.fromLogicalToDataType(rowType)
            );

        final DeltaSink<RowData> deltaSink = DeltaSink
            .forRowData(
                new org.apache.flink.core.fs.Path(deltaTablePath),
                DeltaTestUtils.getHadoopConf(),
                rowType
            ).build();

        final DataStream<RowData> inputStream = env.fromElements(
            typeConverter.toInternal(Row.of(LocalDate.of(2024, 1, 1), 0))
        );

        inputStream.sinkTo(deltaSink);

        env.execute("Delta Sink Example");
    }
}
```

File: connectors/flink/src/test/java/io/delta/flink/sink/DeltaSinkITCase.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+                .mapToObj(startDate::plusDays)
+                .map(date -> Row.of(date, 0))
+                .map(typeConverter::toInternal)
+                .toArray(RowData[]::new);
+
+        final DataStream<RowData> inputStream = env.fromElements(elements);
+
+        inputStream.sinkTo(deltaSink);
+
+        env.execute("Delta Sink Example");
```
Comment (scottsand-db) [inline connectors/flink/src/test/java/io/delta/flink/sink/DeltaSinkITCase.java:58]: next, please validate that the table was written correctly

**Review by scottsand-db — APPROVED** [review]

**Comment by tuliocavalcanti** [comment]
Hi Scott, just double checking, you folks will merge it when its time, right?

**Pull request merged**